ops/jujuversion.py

```diff
@@ -14,133 +14,7 @@
-        if isinstance(other, str):
-            other = type(self)(other)
-        elif not isinstance(other, JujuVersion):
-            raise RuntimeError(f'cannot compare Juju version "{self}" with "{other}"')
-        if self.major != other.major:
-            return self.major < other.major
-        elif self.minor != other.minor:
-            return self.minor < other.minor
-        elif self.tag != other.tag:
-            if not self.tag:
-                return False
-            elif not other.tag:
-                return True
-            return self.tag < other.tag
-        elif self.patch != other.patch:
-            return self.patch < other.patch
-        elif self.build != other.build:
-            return self.build < other.build
-        return False
-
-    @classmethod
-    def from_environ(cls) -> 'JujuVersion':
-        """Build a version from the ``JUJU_VERSION`` environment variable."""
-        v = os.environ.get('JUJU_VERSION')
-        if v is None:
-            v = '0.0.0'
-        return cls(v)
-
-    def has_app_data(self) -> bool:
-        """Report whether this Juju version supports app data."""
-        return (self.major, self.minor, self.patch) >= (2, 7, 0)
-
-    def is_dispatch_aware(self) -> bool:
-        """Report whether this Juju version supports dispatch."""
-        return (self.major, self.minor, self.patch) >= (2, 8, 0)
-
-    def has_controller_storage(self) -> bool:
-        """Report whether this Juju version supports controller-side storage."""
-        return (self.major, self.minor, self.patch) >= (2, 8, 0)
-
-    @property
-    def has_secrets(self) -> bool:
-        """Report whether this Juju version supports the "secrets" feature."""
-        # Juju version 3.0.0 had an initial version of secrets, but:
-        # * In 3.0.2, secret-get "--update" was renamed to "--refresh", and
-        #   secret-get-info was separated into its own hook tool
-        # * In 3.0.3, a bug with observer labels was fixed (juju/juju#14916)
-        return (self.major, self.minor, self.patch) >= (3, 0, 3)
-
-    @property
-    def supports_open_port_on_k8s(self) -> bool:
-        """Report whether this Juju version supports open-port on Kubernetes."""
-        # Support added: https://bugs.launchpad.net/juju/+bug/1920960
-        return (self.major, self.minor, self.patch) >= (3, 0, 3)
-
-    @property
-    def supports_exec_service_context(self) -> bool:
-        """Report whether this Juju version supports exec's service_context option."""
-        if (self.major, self.minor, self.patch) < (3, 1, 6):
-            # First released in 3.1.6
-            return False
-        if (self.major, self.minor, self.patch) == (3, 2, 0):
-            # 3.2.0 was released before Pebble was updated, but all other 3.2
-            # releases have the change (3.2.1 tag was never released).
-            return False
-        return True
+# For backward compatibility only; please use `from ops import JujuVersion`.
```

Comment:
Now that I realize we need to keep this file anyway, I'm having second thoughts about moving JujuVersion to jujucontext.py -- it's confusing. Let's move JujuVersion back here, but keep `_JujuContext` in jujucontext.py. (We would still import `JujuVersion` in jujucontext.py, but not the other way around.)

Sorry to make busy-work!